File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.java
class Solution {
    public int tribonacci(int n) {
        int[] dp = new int[n+1];
        Arrays.fill(dp,-1);
        return helper(n,dp);
    }
    
    private int helper(int n,int[] dp){
        if(n <= 1){  // base case
            return n;
        }
        
        if(n == 2){ // base case;
            return 1;
        }
        
        if(dp[n] != -1){
            return dp[n];
        }
        
        dp[n] =  helper(n-1,dp)+helper(n-2,dp)+helper(n-3,dp);
        return dp[n];
    }
}

File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.cpp
class Solution {
public:
    int tribonacci(int n) {
        if(n == 0 || n == 1){
            return n;
        }
        
        vector<int> dp(n+1,-1);
        dp[0] = 1;
        dp[1] = 1;
        dp[2] = 2;
        
        for(int i = 3; i<n; i++){
            dp[i] = dp[i-1] + dp[i-2] + dp[i-3];
        }
        
        return dp[n-1];
    }
};